File: letstalk/src/models/bootstrap.ts
import Immutable from 'immutable';

import { Cohort } from './cohort';
import { UserPersonalInfo } from './user';
import { Connection } from './connection';

export interface Relationship {
  readonly userId: number;
  readonly userType: number;
  readonly firstName: string;
  readonly lastName: string;
  readonly email: string;
  readonly fbId: string | null;
  readonly fbLink: string | null;
  readonly phoneNumber: string | null;
  readonly cohort: Cohort | null;
  readonly description: string | null;
  readonly matchingState: MatchingState;
}

export function getHumanReadableUserType(userType: number): string {
  switch (userType) {
    case 1:
      return 'Mentor';
    case 2:
      return 'Mentee';
    case 3:
      return 'Asker';
    case 4:
      return 'Answerer';
  }
  return 'Unknown';
}

export enum MatchingState {
  Unknown = 0,
  Unverified,
  Verified,
  Expired,
}

export const USER_STATE_ACCOUNT_CREATED = 'account_created';
export const USER_STATE_ACCOUNT_EMAIL_VERIFIED = 'account_email_verified';
export const USER_STATE_ACCOUNT_HAS_BASIC_INFO = 'account_has_basic_info';
export const USER_STATE_ACCOUNT_SETUP = 'account_setup';
export const USER_STATE_ACCOUNT_MATCHED = 'account_matched';

export type UserState =
  | 'account_created'
  | 'account_email_verified'
  | 'account_has_basic_info'
  | 'account_setup'
  | 'account_matched';

export interface ConnectionRequestWithName extends Connection {
  readonly firstName: string;
  readonly lastName: string;
}

export interface BootstrapConnection {
  readonly userProfile: Relationship;
  readonly request: Connection;
}

interface BootstrapConnections {
  readonly outgoingRequests: Immutable.List<ConnectionRequestWithName>;
  readonly incomingRequests: Immutable.List<ConnectionRequestWithName>;
  readonly mentors: Immutable.List<BootstrapConnection>;
  readonly mentees: Immutable.List<BootstrapConnection>;
  readonly peers: Immutable.List<BootstrapConnection>;
}

export interface BootstrapData {
  readonly state: UserState;
  readonly connections: BootstrapConnections;
};

File: server/core/api/bootstrap.go
package api

import (
	"letstalk/server/data"
)

type BootstrapUserRelationshipDataModel struct {
	UserId        data.TUserID       `json:"userId" binding:"required"`
	UserType      UserType           `json:"userType" binding:"required"`
	FirstName     string             `json:"firstName" binding:"required"`
	LastName      string             `json:"lastName" binding:"required"`
	Email         string             `json:"email" binding:"required"`
	FbId          *string            `json:"fbId"`
	FBLink        *string            `json:"fbLink" binding:"required"`
	PhoneNumber   *string            `json:"phoneNumber"`
	Cohort        *Cohort            `json:"cohort"`
	Description   *string            `json:"description"`
	MatchingState data.MatchingState `json:"matchingState"`
}

type ConnectionRequestWithName struct {
	ConnectionRequest
	FirstName string `json:"firstName" binding:"required"`
	LastName  string `json:"lastName" binding:"required"`
}

type BootstrapConnection struct {
	UserProfile BootstrapUserRelationshipDataModel `json:"userProfile" binding:"required"`
	Request     ConnectionRequest                  `json:"request" binding:"required"`
}

type BootstrapConnections struct {
	OutgoingRequests []*ConnectionRequestWithName `json:"outgoingRequests" binding:"required"`
	IncomingRequests []*ConnectionRequestWithName `json:"incomingRequests" binding:"required"`
	Mentors          []*BootstrapConnection       `json:"mentors" binding:"required"`
	Mentees          []*BootstrapConnection       `json:"mentees" binding:"required"`
	Peers            []*BootstrapConnection       `json:"peers" binding:"required"`
}

type BootstrapResponse struct {
	State       UserState            `json:"state" binding:"required"`
	Connections BootstrapConnections `json:"connections" binding:"required"`
}
